File: architecture-skills/design-patterns/examples/proxy/proxy.cs
// Proxy pattern in C# (virtual proxy: lazy-loading image).
//
// The classic interface + delegating class form. Lazy<T> can replace
// the null check for thread-safe lazy creation, and DispatchProxy
// generates delegation for arbitrary interfaces at runtime (the
// mechanism behind many ORM and mocking libraries).

using System;

public interface IImage
{
    void Display();
}

// Real subject: expensive to create.
public sealed class RealImage : IImage
{
    private readonly string _filename;

    public RealImage(string filename)
    {
        _filename = filename;
        Console.WriteLine($"Loading {filename} from disk (expensive)");
    }

    public void Display() => Console.WriteLine($"Displaying {_filename}");
}

// Virtual proxy: defers creating RealImage until first use.
public sealed class ImageProxy : IImage
{
    private readonly Lazy<RealImage> _real;

    public ImageProxy(string filename)
    {
        _real = new Lazy<RealImage>(() => new RealImage(filename));
    }

    public void Display() => _real.Value.Display();
}

public static class Program
{
    public static void Main()
    {
        IImage holiday = new ImageProxy("holiday.png");
        IImage unused = new ImageProxy("unused.png");
        Console.WriteLine("Proxies created; nothing loaded yet");

        holiday.Display();
        holiday.Display(); // loaded once, displayed twice

        Console.WriteLine("unused.png was never loaded");
    }
}

File: architecture-skills/design-patterns/examples/proxy/proxy.java
// Proxy pattern in Java (virtual proxy: lazy-loading image).
//
// The classic interface + delegating class form. For proxying arbitrary
// interfaces at runtime, java.lang.reflect.Proxy generates the
// delegation dynamically — the mechanism behind most ORM lazy loading
// and mocking frameworks.

public class proxy {

    interface Image {
        void display();
    }

    /** Real subject: expensive to create. */
    static final class RealImage implements Image {
        private final String filename;

        RealImage(String filename) {
            this.filename = filename;
            System.out.println("Loading " + filename + " from disk (expensive)");
        }

        @Override
        public void display() {
            System.out.println("Displaying " + filename);
        }
    }

    /** Virtual proxy: defers creating RealImage until first use. */
    static final class ImageProxy implements Image {
        private final String filename;
        private RealImage real;

        ImageProxy(String filename) {
            this.filename = filename;
        }

        @Override
        public void display() {
            if (real == null) {
                real = new RealImage(filename);
            }
            real.display();
        }
    }

    public static void main(String[] args) {
        Image holiday = new ImageProxy("holiday.png");
        Image unused = new ImageProxy("unused.png");
        System.out.println("Proxies created; nothing loaded yet");

        holiday.display();
        holiday.display(); // loaded once, displayed twice

        System.out.println("unused.png was never loaded");
    }
}
